### DISCUSSION TITLE
wsl.exe periodically sends control characters making impossible "piping" via interactive/terminal mode

### DISCUSSION THREAD

**Discussion by gavenkoa** [microsoft/WSL#6591] [General]
I tried to set WSL 1 connection type for Emacs TRAMP:

https://www.reddit.com/r/emacs/comments/jrkgmy/tramp_for_wsl_1_from_cygwin_emacsw32/

and debugging led me to code:

```
(setq xxx (make-process :name "xxx" :buffer "xxx" :command '("wsl" "echo" "OK") :connection-type 'pty))
(stop-process xxx)
```

which shows following (I replaced occurrences of ESC byte, or #x1b, with `\e` ):

```
\e[25l\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K
\e[K\e[H\e[?25hOK
```

It is not clear if it comes from `conhost.exe` or from `wsl.exe`.

`pipe` mode or `wsl echo OK | cat` disable garbage appearance but unfortunately Emacs TRAMP implementation starts `wsl` in the intermediate step and assumes that there is only one terminal (Cygwin's in my case), who can imagine that WSL creates intermediate "terminal"?

Also if I launch interactive program, like `sh` WSL (or conhost??) **periodically emits similar garbage**. It make impossible to pipe data as they will be corrupted.

So far:

* there is a need for clarification who is responsible for garbage emmition
* if it is `wsl` itself we need special flag to disable terminal "smartness" of WSL - to tell to stop to emit any control sequences to clear terminal or 

Why:

* it is impossible to use external (non-WSL) `expect` -like programs (for example Emacs or `expect` from Cygwin) to automate interaction with WSL 1 via `wsl cmd -opt1 -op2` like invocation.

**Comment by DHowett** [comment]
I believe this is due to the allocation of a pty. those escapes are coming out of conhost _because the process was opted in for an interactive session_. This is like giving a tty to a process: `isatty` returns true, and the application will likely assume it is in an interactive session and use screen and formatting controls.

**Comment by Biswa96** [comment]
Can you try [wsltty](https://github.com/mintty/wsltty) with that same test?

**Comment by gavenkoa** [comment]
> Can you try wsltty with that same test?

It is impossible. The way `wsltty` installs and behaves makes it unusable from Cygwin.

As I understand the goal of this project is to launch Cygwin but not be launched by Cygwin.

**Comment by gavenkoa** [comment]
@DHowett 

> I believe this is due to the allocation of a pty.

That's true.

I tied to trick Cygwin Emacs TRAMP mode by piping WSL over Cygwin's `cat`:

```
'((tramp-login-program "sh")
  (tramp-login-args (("-c" "'wsl -- sh | cat'")))
```

Emacs TRAMP mode sends to Cygwin sh interpreter:

```
23:02:20.075911 tramp-maybe-open-connection (3) # Sending command ‘exec sh -c 'wsl -- sh | cat'’
```

but still WSL generates ESC control sequences.

There is no such issue when I execute `sh -c 'wsl -- sh | cat'` from mintty. 

Special option for activating WSL in PIPE mode will resolve this problem.

I think that some thin custom wrapper that spawns WSL via pipe also will help, my attempts to write Cygwin script with "wsl -- sh | cat" didn't help. It should be written in C or with `expect` trickery...

**Comment by DHowett** [comment]
I don’t know much about emacs, but couldn’t you remove `:connection-type 'pty` to not request a PTY?

**Comment by gavenkoa** [comment]
@DHowett 

> I don’t know much about emacs, but couldn’t you remove :connection-type 'pty to not request a PTY?

The problem is that Emacs TRAMP starts `sh` process and then communicate via pipe / terminal with SH. They send `exec wsl` (or whatever I configure) in an intermediate shell. At this stage I have no control over `wsl` executable (like to force it into a pipe mode). 

When I work directly with Emacs process API and change it to a pipe mode - there is no issues:

```
(setq xxx (make-process :name "xxx" :buffer "xxx" :command '("wsl" "sh") :connection-type 'pipe))
(process-send-string xxx "ls\n")
(stop-process xxx)
```

So it is a specific design of TRAMP (designed in a way similar to Ansible) prevents me to influence how WSL is started. Because they spawn an intermediate process. To make it work one is needed:

* Rewrite TRAMP to avoid launching intermediate shell, and to start WSL directly with `pipe` option
* Add option `--force-pipe` to WSL to  avoid ESC control sequences.
* Write a wrapper that starts WSL in a pipe mode (in that separate process we have full control over WSL).

**Comment by gavenkoa** [comment]
I'll ask for advice in Cygwin and/or Emacs TRAMP mail lists in order for a solution that doesn't require changes on WSL sde.

**Comment by gavenkoa** [comment]
I've got alternative solution that obsoletes need for TRAMP.

After some hints from Cygwin mailing list I managed to "mount" WSL network P9 `\\wsl?\distro` points in Cygwin and have idiomatic way to access files under WSL.

Details are here: https://cygwin.com/pipermail/cygwin/2020-November/246886.html

The original report with suggestion to stop WSL (or conemu??) emitting special clearing sequences still valid as it might break another Expect-like bidirectional communication programs that jump from native Windows into WSL via `wsl.exe -d debian /bin/sh`
